<!DOCTYPE html>
<html>
  
<head>

  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Aligning DNA</title>
  <meta name="description" content="                      Brentgaither.GitHub.io by brentgaither  ">

  <link rel="stylesheet" href="/css/main.css">
  <link rel="canonical" href="http://brentgaither.github.io/2016/02/19/MatchingSequences.html">
  <link rel="alternate" type="application/rss+xml" title="Brent Gaither" href="http://brentgaither.github.io/feed.xml">
</head>


  <body>

    

<header class="site-header">

  <div class="wrapper">

    <a class="site-title" href="/">Brent Gaither</a>

    <nav class="site-nav">
      <a href="#" class="menu-icon">
<!--         <svg viewBox="0 0 18 15">
          <path fill="#424242" d="M18,1.484c0,0.82-0.665,1.484-1.484,1.484H1.484C0.665,2.969,0,2.304,0,1.484l0,0C0,0.665,0.665,0,1.484,0 h15.031C17.335,0,18,0.665,18,1.484L18,1.484z"/>
          <path fill="#424242" d="M18,7.516C18,8.335,17.335,9,16.516,9H1.484C0.665,9,0,8.335,0,7.516l0,0c0-0.82,0.665-1.484,1.484-1.484 h15.031C17.335,6.031,18,6.696,18,7.516L18,7.516z"/>
          <path fill="#424242" d="M18,13.516C18,14.335,17.335,15,16.516,15H1.484C0.665,15,0,14.335,0,13.516l0,0 c0-0.82,0.665-1.484,1.484-1.484h15.031C17.335,12.031,18,12.696,18,13.516L18,13.516z"/>
        </svg> -->
      </a>

      <div class="trigger">
        
          
          <a class="page-link" href="/about/">About</a>
          
        
          
        
          
        
      </div>
    </nav>

  </div>

</header>


    <div class="page-content">
      <div class="wrapper">
        <article class="post" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header">
    <h1 class="post-title" itemprop="name headline">Aligning DNA</h1>
    <p class="post-meta"><time datetime="2016-02-19T00:00:00-07:00" itemprop="datePublished">Feb 19, 2016</time></p>
  </header>

  <div class="post-content" itemprop="articleBody">
    <html>
  <head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <link href="https://fonts.googleapis.com/css?family=Chivo:900" rel="stylesheet" type="text/css" />
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
    <title>Brentgaither.GitHub.io by brentgaither</title>
  </head>

  <body>
    <div id="container">
      <div class="inner">

        <hr />

        <section id="main_content">
          
<p>This first program is here to show how easy it is to start learning about bioinformatics using python. This program
quickly anaylzes two strings or DNA strands to see how similar they are. This even includes a few additional features
for alignment style and direction. This does not however take into account all of the factors that a real DNA strand will encounter.</p>
<p>I started with a simple method to read in a <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> formatted file. This is used in most of my programs since FASTA files are a standard. </p>


<pre class="prettyprint"><code class="language-python">
  
<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="k">def</span> <span class="nf">readInFile</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span> \<span class="n">n</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="s">""</span> 
    <span class="n">infile</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>  <span class="c"># bypass &gt; header line</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">'</span><span class="se">\n</span><span class="s">'</span><span class="p">,</span> <span class="s">''</span><span class="p">)</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span> <span class="o">+</span> <span class="n">line</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">sequence</span>
    </code></pre></figure>

  </code></pre>

  <p>This next method comapres two sequences with the same length. This is a rather crude way to compare DNA sequences but to start to learn how to use python in biology this works. </p>
  <pre class="prettyprint"><code class="language-python">
    
<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="k">def</span> <span class="nf">comparison</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">):</span>
        
    <span class="n">counterComp</span> <span class="o">=</span> <span class="mi">0</span> 

    <span class="k">for</span> <span class="n">iterator</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)):</span>
        <span class="c">#find matching nucleotides and add total </span>
        <span class="k">if</span><span class="p">(</span><span class="n">seq1</span><span class="p">[</span><span class="n">iterator</span><span class="p">]</span> <span class="o">==</span> <span class="n">seq2</span><span class="p">[</span><span class="n">iterator</span><span class="p">]):</span>
             <span class="n">counterComp</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">if</span><span class="p">(</span><span class="n">counterComp</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
        <span class="k">print</span> <span class="p">(</span><span class="s">"no mismatches found - sequences are identical"</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>        
        <span class="k">print</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="s">"Sequences are identical length with "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">counterComp</span><span class="p">)</span> 
            <span class="o">+</span> <span class="s">" matches out of "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">))</span> <span class="o">+</span> <span class="s">" nucleotides"</span><span class="p">)</span>
  </code></pre></figure>

  </code></pre>
  <p> We can now input a FASTA file and compare sequences with this method. What happens if we have sequences of different lengths? We can use this method below to try and predict where the sequences might line up.</p>
  <pre class="prettyprint"><code class="language-python">
    
<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="k">def</span> <span class="nf">comparisonUnequal</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">):</span>
    <span class="n">difference</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">deletion</span> <span class="o">=</span><span class="s">''</span>

    <span class="n">bestFit</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s">''</span><span class="p">)</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)):</span>
        <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">switchSequences</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">)</span>

    <span class="n">difference</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
    <span class="n">shortSeq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>

    <span class="c">#create spaces for deletion</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">difference</span><span class="p">):</span>
        <span class="n">deletion</span> <span class="o">=</span> <span class="n">deletion</span> <span class="o">+</span> <span class="s">'-'</span>   

    <span class="c">#iterate through sequence moving deletion down each nucleotide</span>
    <span class="k">for</span> <span class="n">iterator</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">shortSeq</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#resets counter to 0</span>
        <span class="n">newSeq2</span> <span class="o">=</span> <span class="n">seq2</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">iterator</span><span class="p">]</span> <span class="o">+</span> <span class="n">deletion</span> <span class="o">+</span> <span class="n">seq2</span><span class="p">[</span><span class="n">iterator</span><span class="p">:</span><span class="n">shortSeq</span><span class="p">]</span> 
        <span class="c">#count number of matching nucleotides put in list</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">shortSeq</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">seq1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">newSeq2</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                <span class="n">counter</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">counter</span><span class="p">)</span><span class="o">+</span> <span class="mi">1</span>
        <span class="n">bestFit</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">counter</span><span class="p">))</span>
    <span class="k">print</span><span class="p">(</span><span class="s">"The best matching sequence is..."</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">seq2</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))]</span> <span class="o">+</span> <span class="n">deletion</span> <span class="o">+</span> 
        <span class="n">seq2</span><span class="p">[</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">)):</span><span class="n">shortSeq</span><span class="p">])</span>
    <span class="k">print</span><span class="p">(</span><span class="s">"There are "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))</span> <span class="o">+</span> <span class="s">" matching nucleotides"</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s">"A deletion of "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">difference</span><span class="p">)</span> <span class="o">+</span> 
        <span class="s">" nucleotide(s) occurred at nucleotide(s) "</span> 
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="s">"-"</span><span class="o">+</span> 
        <span class="nb">str</span><span class="p">(</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))</span><span class="o">+</span><span class="n">difference</span><span class="p">))</span>
  </code></pre></figure>

  </code></pre>
  <p> I also added in a method to ensure that I knew which DNA strand was longer. This helps me ensure i have a positive number when determining how many nucleotides are missing or inserted between the two sequences. 
  </p>
  <pre class="prettyprint"><code class="language-python">
    
<figure class="highlight"><pre><code class="language-python" data-lang="python">  <span class="k">def</span> <span class="nf">switchSequences</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">):</span>
      <span class="n">temp</span> <span class="o">=</span> <span class="n">seq1</span>
      <span class="n">seq1</span> <span class="o">=</span> <span class="n">seq2</span>
      <span class="n">seq2</span> <span class="o">=</span> <span class="n">temp</span>
      <span class="k">return</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span>
  </code></pre></figure>

  </code></pre>
        </section>


                  <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-73767757-1");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

      </div>
    </div>
  </body>
</html>

  </div>

</article>

      </div>
    </div>

    <footer class="site-footer">

  <div class="wrapper">

    <h2 class="footer-heading">Brent Gaither</h2>

    <div class="footer-col-wrapper">
      <div class="footer-col footer-col-1">
        <ul class="contact-list">
          <li>Brent Gaither</li>
          <li><a href="mailto:brent[dot]w[dot]gaither[at]gmail[dot]com">brent[dot]w[dot]gaither[at]gmail[dot]com</a></li>
        </ul>
      </div>

      <div class="footer-col footer-col-2">
        <ul class="social-media-list">
          
          <li>
            <!-- <a href="https://github.com/brentgaither"><span class="icon icon--github"><svg viewBox="0 0 16 16"><path fill="#828282" d="M7.999,0.431c-4.285,0-7.76,3.474-7.76,7.761 c0,3.428,2.223,6.337,5.307,7.363c0.388,0.071,0.53-0.168,0.53-0.374c0-0.184-0.007-0.672-0.01-1.32 c-2.159,0.469-2.614-1.04-2.614-1.04c-0.353-0.896-0.862-1.135-0.862-1.135c-0.705-0.481,0.053-0.472,0.053-0.472 c0.779,0.055,1.189,0.8,1.189,0.8c0.692,1.186,1.816,0.843,2.258,0.645c0.071-0.502,0.271-0.843,0.493-1.037 C4.86,11.425,3.049,10.76,3.049,7.786c0-0.847,0.302-1.54,0.799-2.082C3.768,5.507,3.501,4.718,3.924,3.65 c0,0,0.652-0.209,2.134,0.796C6.677,4.273,7.34,4.187,8,4.184c0.659,0.003,1.323,0.089,1.943,0.261 c1.482-1.004,2.132-0.796,2.132-0.796c0.423,1.068,0.157,1.857,0.077,2.054c0.497,0.542,0.798,1.235,0.798,2.082 c0,2.981-1.814,3.637-3.543,3.829c0.279,0.24,0.527,0.713,0.527,1.437c0,1.037-0.01,1.874-0.01,2.129 c0,0.208,0.14,0.449,0.534,0.373c3.081-1.028,5.302-3.935,5.302-7.362C15.76,3.906,12.285,0.431,7.999,0.431z"/></svg>
</span><span class="username">brentgaither</span></a>
 -->
          </li>
          

        </ul>
        <footer>
          <br><br>
          <img src="/images/octocat-icon.png"><a href="https://pages.github.com">GitHub Pages</a>.
        </footer>
      </div>

      <div class="footer-col footer-col-3">
        <p>Brent Gaither. This is my personal site for bioinformatics and other cool things I am learning about. 
</p>
      </div>
    </div>

  </div>

</footer>


  </body>

</html>
