<!DOCTYPE html>
<html>
  
<head>

  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1">

  <title>Aligning DNA</title>
  <meta name="description" content="">

  <link rel="stylesheet" href="/css/main.css">
  <link rel="stylesheet" href="/css/bootstrap.min.css">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.2/jquery.min.js"></script>
  
  <link rel="canonical" href="http://brentgaither.github.io/2016/02/19/MatchingSequences.html">
  <link rel="alternate" type="application/rss+xml" title="Brent Gaither" href="http://brentgaither.github.io/feed.xml">
</head>


  <nav class="navbar navbar-default">
  <div class="container-fluid col-md-12 col-lg-12">
    <span role="presentation" class="active"><a href="/">Brent Gaither</a></span>    
    <span class="col-lg-offset-11" role="presentation"><a href="/about">About</a></span>
    <span class="col-lg-offset-11" role="presentation"><a href="/PokemonGo">PokemonGo</a></span>
  </div>
</nav>


  <body>

   

    <div class="page-content">
        <article class="post" itemscope itemtype="http://schema.org/BlogPosting">

  <header class="post-header col-md-offset-1 col-lg-offset-1">
    <h1 class="post-title" itemprop="name headline">Aligning DNA</h1>
    <p class="post-meta"><time datetime="2016-02-19T00:00:00-07:00" itemprop="datePublished">Feb 19, 2016</time></p>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <link href='https://fonts.googleapis.com/css?family=Chivo:900' rel='stylesheet' type='text/css'>
    <title>Brentgaither.GitHub.io by brentgaither</title>
  </header>

  <div class="post-content" itemprop="articleBody">
    <div class="col-lg-12">
<div class="col-lg-7 col-lg-offset-2">
    <html>


<body>
  <div id="container">
    <div class="inner">
      <hr />  

<section id="main_content">
          

<p>This first program is here to show how easy it is to start learning about bioinformatics using python. This program
quickly anaylzes two strings or DNA strands to see how similar they are. This even includes a few additional features
for alignment style and direction. This does not however take into account all of the factors that a real DNA strand will encounter.</p>
<p>I started with a simple method to read in a <a href="https://en.wikipedia.org/wiki/FASTA_format">FASTA</a> formatted file. This is used in most of my programs since FASTA files are a standard. </p>



  
<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="k">def</span> <span class="nf">readInFile</span><span class="p">(</span><span class="n">infile</span><span class="p">):</span> \<span class="n">n</span>
    <span class="n">sequence</span> <span class="o">=</span> <span class="s">""</span> 
    <span class="n">infile</span><span class="o">.</span><span class="n">readline</span><span class="p">()</span>  <span class="c"># bypass &gt; header line</span>
    <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">infile</span><span class="p">:</span>
        <span class="n">line</span> <span class="o">=</span> <span class="n">line</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s">'</span><span class="se">\n</span><span class="s">'</span><span class="p">,</span> <span class="s">''</span><span class="p">)</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span> <span class="o">+</span> <span class="n">line</span>
        <span class="n">sequence</span> <span class="o">=</span> <span class="n">sequence</span><span class="o">.</span><span class="n">upper</span><span class="p">()</span>
    <span class="n">infile</span><span class="o">.</span><span class="n">close</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">sequence</span>
    </code></pre></figure>



  <p>This next method comapres two sequences with the same length. This is a rather crude way to compare DNA sequences but to start to learn how to use python in biology this works. </p>

    
<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="k">def</span> <span class="nf">comparison</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">):</span>
        
    <span class="n">counterComp</span> <span class="o">=</span> <span class="mi">0</span> 

    <span class="k">for</span> <span class="n">iterator</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)):</span>
        <span class="c">#find matching nucleotides and add total </span>
        <span class="k">if</span><span class="p">(</span><span class="n">seq1</span><span class="p">[</span><span class="n">iterator</span><span class="p">]</span> <span class="o">==</span> <span class="n">seq2</span><span class="p">[</span><span class="n">iterator</span><span class="p">]):</span>
             <span class="n">counterComp</span> <span class="o">+=</span> <span class="mi">1</span>
    <span class="k">if</span><span class="p">(</span><span class="n">counterComp</span> <span class="o">==</span> <span class="mi">0</span><span class="p">):</span>
        <span class="k">print</span> <span class="p">(</span><span class="s">"no mismatches found - sequences are identical"</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>        
        <span class="k">print</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
        <span class="k">print</span><span class="p">(</span><span class="s">"Sequences are identical length with "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">counterComp</span><span class="p">)</span> 
            <span class="o">+</span> <span class="s">" matches out of "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">))</span> <span class="o">+</span> <span class="s">" nucleotides"</span><span class="p">)</span>
  </code></pre></figure>


  <p> We can now input a FASTA file and compare sequences with this method. What happens if we have sequences of different lengths? We can use this method below to try and predict where the sequences might line up.</p>

    
<figure class="highlight"><pre><code class="language-python" data-lang="python"><span class="k">def</span> <span class="nf">comparisonUnequal</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">):</span>
    <span class="n">difference</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">deletion</span> <span class="o">=</span><span class="s">''</span>

    <span class="n">bestFit</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="s">''</span><span class="p">)</span>
    <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">if</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">&lt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)):</span>
        <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span> <span class="o">=</span> <span class="n">switchSequences</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">)</span>

    <span class="n">difference</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span> <span class="o">-</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>
    <span class="n">shortSeq</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq2</span><span class="p">)</span>

    <span class="c">#create spaces for deletion</span>
    <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">difference</span><span class="p">):</span>
        <span class="n">deletion</span> <span class="o">=</span> <span class="n">deletion</span> <span class="o">+</span> <span class="s">'-'</span>   

    <span class="c">#iterate through sequence moving deletion down each nucleotide</span>
    <span class="k">for</span> <span class="n">iterator</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">shortSeq</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
        <span class="n">counter</span> <span class="o">=</span> <span class="mi">0</span> <span class="c">#resets counter to 0</span>
        <span class="n">newSeq2</span> <span class="o">=</span> <span class="n">seq2</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">iterator</span><span class="p">]</span> <span class="o">+</span> <span class="n">deletion</span> <span class="o">+</span> <span class="n">seq2</span><span class="p">[</span><span class="n">iterator</span><span class="p">:</span><span class="n">shortSeq</span><span class="p">]</span> 
        <span class="c">#count number of matching nucleotides put in list</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="n">shortSeq</span><span class="o">+</span><span class="mi">1</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">seq1</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="n">newSeq2</span><span class="p">[</span><span class="n">i</span><span class="p">]:</span>
                <span class="n">counter</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">counter</span><span class="p">)</span><span class="o">+</span> <span class="mi">1</span>
        <span class="n">bestFit</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">int</span><span class="p">(</span><span class="n">counter</span><span class="p">))</span>
    <span class="k">print</span><span class="p">(</span><span class="s">"The best matching sequence is..."</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">seq1</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="n">seq2</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))]</span> <span class="o">+</span> <span class="n">deletion</span> <span class="o">+</span> 
        <span class="n">seq2</span><span class="p">[</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">)):</span><span class="n">shortSeq</span><span class="p">])</span>
    <span class="k">print</span><span class="p">(</span><span class="s">"There are "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))</span> <span class="o">+</span> <span class="s">" matching nucleotides"</span><span class="p">)</span>
    <span class="k">print</span><span class="p">(</span><span class="s">"A deletion of "</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">difference</span><span class="p">)</span> <span class="o">+</span> 
        <span class="s">" nucleotide(s) occurred at nucleotide(s) "</span> 
        <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))</span><span class="o">+</span><span class="mi">1</span><span class="p">)</span> <span class="o">+</span> <span class="s">"-"</span><span class="o">+</span> 
        <span class="nb">str</span><span class="p">(</span><span class="n">bestFit</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="nb">max</span><span class="p">(</span><span class="n">bestFit</span><span class="p">))</span><span class="o">+</span><span class="n">difference</span><span class="p">))</span>
  </code></pre></figure>


  <p> I also added in a method to ensure that I knew which DNA strand was longer. This helps me ensure i have a positive number when determining how many nucleotides are missing or inserted between the two sequences. 
  </p>

    
<figure class="highlight"><pre><code class="language-python" data-lang="python">  <span class="k">def</span> <span class="nf">switchSequences</span><span class="p">(</span><span class="n">seq1</span><span class="p">,</span><span class="n">seq2</span><span class="p">):</span>
      <span class="n">temp</span> <span class="o">=</span> <span class="n">seq1</span>
      <span class="n">seq1</span> <span class="o">=</span> <span class="n">seq2</span>
      <span class="n">seq2</span> <span class="o">=</span> <span class="n">temp</span>
      <span class="k">return</span> <span class="n">seq1</span><span class="p">,</span> <span class="n">seq2</span>
  </code></pre></figure>


  </section>
  </div>
</div>

    </body>
</html>

  </div>
</div>
</div>
</article>

      </div>


    
<div class="vert-offset"></div>
<footer class="site-footer">

  <div class="wrapper col-md-12">

    
        <table class="contact-list col-md-9">
          <tr><td>Brent Gaither<td></tr>
          <tr><td>brent[dot]w[dot]gaither[at]gmail[dot]com<td></tr>
          <tr><td><img src="/images/octocat-icon.png"><a href="https://pages.github.com">GitHub Pages</a></td></tr>
        </table>
      <div class="footer-col col-md-3">
        <p>Brent Gaither. This is my personal site for some interesting problems I have ran into. 
</p>
      </div>

  </div>

</footer>


  </body>

</html>
